File: backend/app/services/pattern_scorer.py
"""
Pattern Scorer — analyses recent 1-min candles for repeating chart patterns,
scores their directional bias, and produces an adjustment signal that the
Kronos 1-min predictor uses to bias its next-candle forecast.

Pattern categories scored:
 1. Classic candlestick patterns (doji, hammer, engulfing, etc.)
 2. Micro-structure repeats  — sliding-window DTW on last N candles
 3. Support / Resistance proximity
 4. Momentum regime (RSI / MACD micro)
"""

import numpy as np
import pandas as pd
import logging
from typing import Optional

logger = logging.getLogger(__name__)


# ── Micro-pattern repeat detector ────────────────────────────────
def _normalise_window(closes: np.ndarray) -> np.ndarray:
    """Zero-mean, unit-variance normalisation of a close series."""
    std = np.std(closes)
    if std < 1e-9:
        return np.zeros_like(closes)
    return (closes - np.mean(closes)) / std


def find_repeating_patterns(
    closes: np.ndarray,
    window: int = 10,
    search_depth: int = 120,
    top_k: int = 3,
    similarity_threshold: float = 0.80,
) -> list[dict]:
    """
    Slide a window of `window` bars back through the last `search_depth` bars
    and find the top-k most similar subsequences to the most-recent window.
    For each match, record what happened in the bar immediately *after* the
    pattern completed — that's our signal for the next candle.
    """
    if len(closes) < window + 2:
        return []

    query = _normalise_window(closes[-window:])
    search_start = max(0, len(closes) - search_depth - window)
    search_end = len(closes) - window - 1  # leave room for "what-came-next"

    matches = []
    for i in range(search_start, search_end):
        candidate = _normalise_window(closes[i : i + window])
        # Pearson correlation as similarity
        corr = float(np.corrcoef(query, candidate)[0, 1])
        if corr >= similarity_threshold:
            next_idx = i + window
            if next_idx < len(closes):
                prev_close = closes[i + window - 1]
                next_close = closes[next_idx]
                pct_change = (next_close - prev_close) / prev_close * 100
                matches.append({
                    "offset": i,
                    "similarity": round(corr, 4),
                    "next_change_pct": round(pct_change, 4),
                })

    # Sort by similarity desc, take top k
    matches.sort(key=lambda m: m["similarity"], reverse=True)
    return matches[:top_k]


# ── Support / Resistance micro-levels ────────────────────────────
def detect_micro_sr(
    highs: np.ndarray, lows: np.ndarray, closes: np.ndarray, lookback: int = 60
) -> dict:
    """Detect the nearest support and resistance from recent pivot points."""
    h = highs[-lookback:]
    l = lows[-lookback:]
    current = closes[-1]

    # Simple pivot: local max/min over 5-bar windows
    resistances = []
    supports = []
    for i in range(2, len(h) - 2):
        if h[i] > h[i - 1] and h[i] > h[i - 2] and h[i] > h[i + 1] and h[i] > h[i + 2]:
            resistances.append(float(h[i]))
        if l[i] < l[i - 1] and l[i] < l[i - 2] and l[i] < l[i + 1] and l[i] < l[i + 2]:
            supports.append(float(l[i]))

    nearest_resistance = min(resistances, key=lambda r: r - current, default=None) if resistances else None
    nearest_support = max(
        [s for s in supports if s < current], key=lambda s: s, default=None
    ) if supports else None

    return {
        "current": float(current),
        "nearest_resistance": nearest_resistance,
        "nearest_support": nearest_support,
        "resistance_distance_pct": round((nearest_resistance - current) / current * 100, 3) if nearest_resistance else None,
        "support_distance_pct": round((current - nearest_support) / current * 100, 3) if nearest_support else None,
    }


# ── Momentum micro-regime ────────────────────────────────────────
def micro_momentum(closes: np.ndarray, n_rsi: int = 14) -> dict:
    """Compute micro RSI and MACD on the 1-min series."""
    if len(closes) < n_rsi + 1:
        return {"rsi": 50.0, "macd_hist": 0.0, "regime": "neutral"}

    deltas = np.diff(closes)
    gains = np.where(deltas > 0, deltas, 0.0)
    losses = np.where(deltas < 0, -deltas, 0.0)

    avg_gain = np.mean(gains[-n_rsi:])
    avg_loss = np.mean(losses[-n_rsi:])
    rs = avg_gain / (avg_loss + 1e-9)
    rsi = 100 - 100 / (1 + rs)

    # Simple MACD: 12-bar EMA - 26-bar EMA
    if len(closes) >= 26:
        ema12 = pd.Series(closes).ewm(span=12).mean().iloc[-1]
        ema26 = pd.Series(closes).ewm(span=26).mean().iloc[-1]
        macd = ema12 - ema26
        signal = pd.Series(closes).ewm(span=12).mean().ewm(span=9).mean().iloc[-1]
        hist = macd - signal
    else:
        hist = 0.0

    if rsi > 70:
        regime = "overbought"
    elif rsi < 30:
        regime = "oversold"
    elif hist > 0:
        regime = "bullish"
    elif hist < 0:
        regime = "bearish"
    else:
        regime = "neutral"

    return {"rsi": round(float(rsi), 2), "macd_hist": round(float(hist), 4), "regime": regime}


# ── Aggregate score ──────────────────────────────────────────────
def compute_pattern_score(df: pd.DataFrame) -> dict:
    """
    Given a DataFrame of 1-min OHLCV candles, compute a combined pattern
    score in [-1.0, +1.0] where +1 is strongly bullish, -1 strongly bearish.
    Also returns component details for the UI.
    """
    closes = df["Close"].values.astype(float)
    highs = df["High"].values.astype(float)
    lows = df["Low"].values.astype(float)

    # 1) Repeating pattern matches
    repeats = find_repeating_patterns(closes, window=10, search_depth=120, top_k=5)
    if repeats:
        avg_next_change = np.mean([m["next_change_pct"] for m in repeats])
        repeat_signal = np.clip(avg_next_change / 0.5, -1, 1)  # normalize
    else:
        avg_next_change = 0.0
        repeat_signal = 0.0

    # 2) Support / Resistance proximity
    sr = detect_micro_sr(highs, lows, closes)
    sr_signal = 0.0
    if sr["resistance_distance_pct"] is not None and sr["resistance_distance_pct"] < 0.1:
        sr_signal = -0.3  # near resistance, bearish bias
    if sr["support_distance_pct"] is not None and sr["support_distance_pct"] < 0.1:
        sr_signal = 0.3  # near support, bullish bias

    # 3) Momentum regime
    mom = micro_momentum(closes)
    mom_signal = 0.0
    if mom["regime"] == "overbought":
        mom_signal = -0.3
    elif mom["regime"] == "oversold":
        mom_signal = 0.3
    elif mom["regime"] == "bullish":
        mom_signal = 0.2
    elif mom["regime"] == "bearish":
        mom_signal = -0.2

    # 4) Recent candle patterns (last 5 bars micro-patterns)
    from app.services.pattern_recognition import detect_all_patterns
    micro_df = df.tail(20).reset_index(drop=True)
    candle_patterns = detect_all_patterns(micro_df)
    candle_signal = 0.0
    if candle_patterns:
        # Weight recent patterns more
        for p in candle_patterns[-5:]:
            if p["type"] == "bullish":
                candle_signal += 0.1 * p.get("confidence", 0.5)
            elif p["type"] == "bearish":
                candle_signal -= 0.1 * p.get("confidence", 0.5)
        candle_signal = np.clip(candle_signal, -1, 1)

    # Weighted combination
    combined = (
        0.40 * repeat_signal +
        0.20 * sr_signal +
        0.20 * mom_signal +
        0.20 * candle_signal
    )
    combined = float(np.clip(combined, -1, 1))

    return {
        "score": round(combined, 4),
        "direction": "bullish" if combined > 0.05 else "bearish" if combined < -0.05 else "neutral",
        "components": {
            "repeat_pattern": {
                "signal": round(float(repeat_signal), 4),
                "matches_found": len(repeats),
                "avg_next_change_pct": round(avg_next_change, 4),
                "matches": repeats,
            },
            "support_resistance": {
                "signal": round(sr_signal, 4),
                **sr,
            },
            "momentum": {
                "signal": round(mom_signal, 4),
                **mom,
            },
            "candle_patterns": {
                "signal": round(float(candle_signal), 4),
                "patterns_detected": [p["name"] for p in candle_patterns[-5:]] if candle_patterns else [],
            },
        },
    }

File: backend-rust/src/services/pattern_scorer.rs
//! Aggregate pattern scoring: repeating-pattern detection, support/resistance,
//! micro-momentum, and candlestick pattern signals.

use crate::models::Candle;
use crate::services::pattern_recognition::{self, PatternResult};
use crate::services::technical;
use serde::Serialize;

#[derive(Debug, Clone, Serialize)]
pub struct PatternScore {
    pub signal: f64,          // -1.0 to 1.0
    pub direction: String,    // "bullish", "bearish", "neutral"
    pub confidence: f64,      // 0.0 to 1.0
    pub repeat_score: f64,
    pub sr_score: f64,
    pub momentum_score: f64,
    pub candle_score: f64,
    pub patterns_found: Vec<PatternResult>,
}

/// Compute the aggregate pattern score for a candle series.
pub fn compute(candles: &[Candle]) -> PatternScore {
    if candles.len() < 10 {
        return PatternScore {
            signal: 0.0,
            direction: "neutral".into(),
            confidence: 0.0,
            repeat_score: 0.0,
            sr_score: 0.0,
            momentum_score: 0.0,
            candle_score: 0.0,
            patterns_found: Vec::new(),
        };
    }

    // 1. Candlestick patterns — focus on recent (last 20 candles)
    let recent_start = candles.len().saturating_sub(20);
    let recent = &candles[recent_start..];
    let patterns = pattern_recognition::detect_all(recent);

    let candle_score = if patterns.is_empty() {
        0.0
    } else {
        let total: f64 = patterns
            .iter()
            .map(|p| p.direction.signal() * p.confidence)
            .sum();
        (total / patterns.len() as f64).clamp(-1.0, 1.0)
    };

    // 2. Repeating pattern detection (Pearson correlation on sliding windows)
    let repeat_score = find_repeating_patterns(candles);

    // 3. Support/Resistance proximity
    let sr_score = detect_sr_signal(candles);

    // 4. Micro-momentum
    let momentum_score = technical::micro_momentum(candles, 5);

    // Weighted combination — require more candle data to reduce noise
    let min_candles_bonus = if candles.len() >= 120 { 1.0 }
        else if candles.len() >= 60 { 0.7 }
        else { 0.4 };

    let signal = (0.40 * repeat_score
        + 0.20 * sr_score
        + 0.20 * momentum_score
        + 0.20 * candle_score)
        .clamp(-1.0, 1.0) * min_candles_bonus;

    // Higher threshold to avoid noise: was 0.1, now 0.2
    let direction = if signal > 0.2 {
        "bullish"
    } else if signal < -0.2 {
        "bearish"
    } else {
        "neutral"
    }
    .into();

    let confidence = signal.abs().min(1.0);

    PatternScore {
        signal,
        direction,
        confidence,
        repeat_score,
        sr_score,
        momentum_score,
        candle_score,
        patterns_found: patterns,
    }
}

/// Sliding-window Pearson correlation to detect repeating price patterns.
fn find_repeating_patterns(candles: &[Candle]) -> f64 {
    if candles.len() < 60 {
        return 0.0;
    }

    let closes: Vec<f64> = candles.iter().map(|c| c.close).collect();
    let n = closes.len();
    let window = 20.min(n / 3);
    let target = &closes[n - window..];
    let target_mean: f64 = target.iter().sum::<f64>() / window as f64;

    let mut best_corr: f64 = 0.0;
    let mut best_next_dir: f64 = 0.0;

    // Scan for matching windows
    let scan_end = n - window - 1;
    let scan_start = scan_end.saturating_sub(200);

    for start in scan_start..scan_end {
        let candidate = &closes[start..start + window];
        let cand_mean: f64 = candidate.iter().sum::<f64>() / window as f64;

        // Pearson correlation
        let mut num = 0.0;
        let mut denom_a = 0.0;
        let mut denom_b = 0.0;
        for j in 0..window {
            let a = target[j] - target_mean;
            let b = candidate[j] - cand_mean;
            num += a * b;
            denom_a += a * a;
            denom_b += b * b;
        }
        let denom = (denom_a * denom_b).sqrt();
        if denom < 1e-10 {
            continue;
        }
        let corr = num / denom;

        if corr > best_corr && start + window < n {
            best_corr = corr;
            // What happened AFTER this matching pattern?
            let after_price = closes[(start + window).min(n - 1)];
            let pattern_end = closes[start + window - 1];
            best_next_dir = if after_price > pattern_end {
                1.0
            } else {
                -1.0
            };
        }
    }

    if best_corr > 0.7 {
        best_next_dir * (best_corr - 0.7) / 0.3 // Scale 0.7-1.0 → 0.0-1.0
    } else {
        0.0
    }
}

/// Detect support/resistance levels and return signal based on proximity.
fn detect_sr_signal(candles: &[Candle]) -> f64 {
    if candles.len() < 20 {
        return 0.0;
    }

    let n = candles.len();
    let current = candles[n - 1].close;
    let window = 5;

    // Find pivot highs and lows
    let mut supports = Vec::new();
    let mut resistances = Vec::new();

    for i in window..(n - window) {
        let is_pivot_high = (0..window).all(|j| {
            candles[i].high >= candles[i - j - 1].high && candles[i].high >= candles[i + j + 1].high
        });
        let is_pivot_low = (0..window).all(|j| {
            candles[i].low <= candles[i - j - 1].low && candles[i].low <= candles[i + j + 1].low
        });

        if is_pivot_high {
            resistances.push(candles[i].high);
        }
        if is_pivot_low {
            supports.push(candles[i].low);
        }
    }

    if supports.is_empty() && resistances.is_empty() {
        return 0.0;
    }

    let atr = candles[n - 1].range(); // Approximate

    // Near support → bullish, near resistance → bearish
    let near_support = supports.iter().any(|s| (current - s).abs() < atr * 0.5);
    let near_resistance = resistances.iter().any(|r| (current - r).abs() < atr * 0.5);

    match (near_support, near_resistance) {
        (true, false) => 0.5,
        (false, true) => -0.5,
        _ => 0.0,
    }
}
